### ISSUE TITLE
Sign up - Verify email - Log in with Auth0

### ISSUE THREAD

**Issue by janagombitova** [akvo/akvo-platform#154]
# Content 
We have decided to move to Auth0 to handle user authentication for Flow and Lumen. Users will be able to login using Google email accounts (where Auth0 acts as a proxy) or any email of their preference. 

With this change, we need to create a way for users to access Flow and Lumen using accounts that are not Google accounts. Thus Auth0 not only knows this user exists but also holds her password. 

## Why do we add this issue 
We want to make sure that the Auth0 change is:

* as smooth as possible in terms of as least possible obstruction to existing users as we can make with sufficient communication
* unifies Flow and Lumen a bit more in terms of user accounts, look and feel and workflows
* we do not remove the important features out of the blue
* we keep the changes as simple and minimal as possible to be able to improve the UX in later iterations

# Solution & next steps
## Status quo 
Currently, in Flow, there is no way to log in not using a Google account, so this functionality does not exist at the moment. 

In Lumen, user accounts live in Keycloak. But as we are moving away from Keycloak we need to make signing up to Auth0 possible. 

## User workflow 
0. Organisational admin adds Salim to a Flow/Lumen instance and tells him which email he needs to use to access and which URL to go to. 
1. Salim goes to the given URL and sees the login page. 
2. As he does not have a password yet, he clicks to `sign up`. 
3. He types in the email and creates a password and hits the button to sign up. 
4. Now he sees a page that tells him to verify his email. 
5. He goes to his email to verify it by clicking on the given link (I assume) 
6. He now needs to log in using this email and the password he created. 

## Next steps  
* [x] Create the email verification workflow and page mockups and UI designs 
* [x] Define how the email should look like and what it should say
* [x] Check if Auth0 can handle the workflow itself, so that instead of redirecting to Flow and/or Lumen after the sign-up, it redirects to itself or a generic page with the user instructions
* [x] If no, we need to implement the same page and logic in both Flow and Lumen (issues to be created if that is the case)

**Comment by Kiarii** [comment]
Here is a [wireframe prototype](https://invis.io/HPT684SSABV#/375439063_0-_A0_-_Lander) depicting the preliminary workflow on Auth0 for: 

- Login, (Lumen language change) and logout 
- Sign up 
- Password reset. 

We still need to look into the phrasing of the screens incl. the (verify email to activate account and password reset) emails

**Comment by janagombitova** [comment]
Today I checked how the **signup** process works and what we are still missing. Kiarii did provide mock-ups for the workflow, but we learned that without hosting Auth0 login page ourselves, there is minimum customisation we can do. With this in mind, we need to figure out how to make the signup process smooth and understandable with the Auht0 limitations in place. 

 ## Current workflow

1. Click to sign up to Auth0
2. Add email and password and hit to sign up
3. Screen - needs branding and love - telling about the need to verify email before logging in to: [ url ] + email sent to the user 

![Screen Shot 2019-08-29 at 14 43 09](https://user-images.githubusercontent.com/12456965/63941706-72a1dd00-ca6c-11e9-97e9-d923db727937.png)

4. After clicking on the verify link in the email, the user is brought to a success screen. 
![Screen Shot 2019-08-29 at 14 44 19](https://user-images.githubusercontent.com/12456965/63941728-79c8eb00-ca6c-11e9-8706-7db8b333bed6.png)


5. Now the user should go the the Lumen/Flow instance but there is no direct way to get there. Once going to the url, the user is logged in. 

## Issues 
Overall the workflow is good but there are some details we need to tweak: 
- ensure the Verify your email screen follows the branding we set up @loicsans can you take a look at this? 
- once I am in the screen that tells me my email is successfully verified, I have no way to go to Lumen/Flow --> can we change this? @tangrammer, @kardan and @loicsans 
- does this need to be implemented per product?

**Comment by tangrammer** [comment]
@janagombitova 

## Issues 
Overall the workflow is good but there are some details we need to tweak: 
- [ ] ensure the Verify your email screen follows the branding we set up @loicsans can you take a look at this? 

after some research we can brand the verification or reset-password screen result with https://auth0.com/docs/email/templates#configuring-the-redirect-to-url var

- [ ] once I am in the screen that tells me my email is successfully verified, I have no way to go to Lumen/Flow --> can we change this? @tangrammer, @kardan and @loicsans 
same as previous, IMO & reading the available options, we only could redirect to a generic url not pre preduct one

- [ ] does this need to be implemented per product? 
see previous comment

- [ ] the email sent comes from an auth0 account. Do we need to and can we configure it to be sent from an akvo email, like noreply@akvo.org or something like that? @tangrammer and @dlebrero 
![Screen Shot 2019-08-29 at 14 55 19](https://user-images.githubusercontent.com/12456965/63941983-125f6b00-ca6d-11e9-889d-6f29fc726c82.png)
we can use our akvo email address

**Comment by tangrammer** [comment]
we need to style the verification error on lumen

<img width="337" alt="Screen Shot 2019-08-29 at 17 43 16" src="https://user-images.githubusercontent.com/731829/63955169-9c66fe00-ca84-11e9-8ce4-0c12c54547d1.png">

**Comment by janagombitova** [comment]
@tangrammer based on your last two comments I suggest to keep this 1st implementation simple and only: 
* [ ] Style the verification error page to follow the overall Auth0 pages branding - Juan, can you coordinate this with Loic? Thanks
* [ ] Use out akvo email address for all the emails

**Comment by tangrammer** [comment]
about styling the verification result page, if we use redirec_url the we'll have following 2 expected page results:
- success redirected_url?supportSignUp=true&supportForgotPassword=true&email=juan%40akvo.org&message=Your%20email%20was%20verified.%20You%20can%20continue%20using%20the%20application.&success=true&code=success#

- fail: redirected_url?email=juan%40akvo.org&message=This%20URL%20can%20be%20used%20only%20once&success=false#

**Comment by janagombitova** [comment]
I tested the workflow again this morning on umen.akvotest.org/library?auth=auth0 and found a few issues: 

* in the verify email the Akvo logo does not load 
   *<img width="1212" alt="Screen Shot 2019-09-05 at 09 00 58" src="https://user-images.githubusercontent.com/12456965/64319275-3adeec00-cfbc-11e9-937a-b79e1576c26a.png">

* the verify email page is still pending 
* after I clicked on the verify email url I got this error page: 
   * <img width="1272" alt="Screen Shot 2019-09-05 at 09 01 30" src="https://user-images.githubusercontent.com/12456965/64319319-55b16080-cfbc-11e9-8629-58cd012ef50b.png">

**Comment by loicsans** [comment]
I've updated the logo image link with a png instead of svg on all email template.

**Comment by tangrammer** [comment]
@kardan updated https://storage.googleapis.com/akvo-public/verify_email.html prototype and it's already added to auth0 settings

**Comment by tangrammer** [comment]
-  [ ] still we should have this page under  `akvo.org` domain ... something like https://xxx.akvo.org/xxx/verify_email.html

**Comment by loicsans** [comment]
I've comited the static branded page here: https://github.com/akvo/public-bucket/blob/master/public/verify_email.html

**Comment by Kiarii** [comment]
I was invited to do a test drive on this, here are some notes:
#### Login and Sign up screens
- there is a "Welcome" text - on both login and signup screens - which am unsure of its purpose as well - this I think is not a necessary UI component to have 
- I find the text on the login screen a bit confusing; "Log in to "workspace name" to continue to Akvo". - "Login into your workspace" or sth less confusing might be better
- The button that should be labeled "Login" is labeled "Continue". This is also applied to the Google button which should read "Login with Google". These also applies to the sign up screens, respectively these should be labeled "Sign up" and "Sign up with Google"; the latter should be centered but its icon left aligned
- The AUTH0 Logo is also shown at the bottom, and links to the Auth0 page, lets ditch it if possible. 
<img width="406" alt="Screenshot 2019-10-03 at 11 03 27" src="https://user-images.githubusercontent.com/12692444/66110541-a45c1580-e5cf-11e9-98db-bf9869427713.png">
<img width="410" alt="Screenshot 2019-10-03 at 11 18 08" src="https://user-images.githubusercontent.com/12692444/66110560-b0e06e00-e5cf-11e9-99dd-7d5d034600d3.png">

**Comment by Kiarii** [comment]
#### Edit profile in lumen
- perhaps this doesnt belong here, but i mention it nonetheless. It should visually appear and behave like a pop-up; atm it appears like an inscreen element instead and doesn't behave as a pop up: clicking outside it does not close it for instance
<img width="264" alt="Screenshot 2019-10-02 at 16 58 57" src="https://user-images.githubusercontent.com/12692444/66110811-36641e00-e5d0-11e9-8ecc-3653b74f1601.png">

**Comment by Kiarii** [comment]
#### Verifying email to activate account
- Here we use a screen title - "Yikes" with a negative connotation while neither the use nor the system has done anything out of the ordinary.. If we have to have a screen title here it ought to be a positive message linked with what the user should do next, not some random..
<img width="425" alt="Screenshot 2019-10-02 at 17 03 16" src="https://user-images.githubusercontent.com/12692444/66110961-817e3100-e5d0-11e9-8875-9ec7203caf78.png">

**Comment by Kiarii** [comment]
#### Terminology discrepancies in ac activation workflow
- **[Account activation email sent screen]** After clicking "continue" on the Sign up screen you get to the _email activation screen_  where the text reads “You_ need to verify your email address. Please check your inbox and click the verification link” - I think the user needs to activate their account by verifying their email. So if we would label this screen "Activate your Akvo account" for instance, the desc text should provide: "An account activation email has been sent to you.." or sth more direct        
- **[email]**, here there is also some discrepancies: The action link reads "Confirm my account" and the preceding text "verify your email". 
- The "Welcome to Akvo" text should be more direct to the task at hand, perhaps, "Activate your Akvo account" 
- The link should also be labeled respectively; "Activate account" perhaps in place of "Confirm..." 
<img width="603" alt="Screenshot 2019-10-03 at 11 44 05" src="https://user-images.githubusercontent.com/12692444/66112230-27329f80-e5d3-11e9-87d8-fa29845a3283.png">

- **[on clicking “confirm my account”]**  [on browser] the user gets “Email verified. Your email address was successfully verified.” - we have to take note that it is the account that was being activated, thus perhaps this should read "Account activated" - HERE - we should provide the user with a way to go to their workspace; else, they are caught up in no man's land and they have to figure out how to go from there to the login screen. A link "Login to Akvo" should thus be inorder.

**Comment by Kiarii** [comment]
#### Password reset screen
- Here as in the Sign up and Login screens we have a button "Continue" - if users dont read, the title "Forgot your password?", an "email field" and a "Continue" button, stand out the most - in which case, we imply to the user that they continue to forget their password :)  
<img width="406" alt="Screenshot 2019-10-03 at 11 55 59" src="https://user-images.githubusercontent.com/12692444/66114201-02402b80-e5d7-11e9-9ee7-4c604f09b560.png">

- A title "Reset password" might be more meaningful. The desc text can read "Forgot password? Enter your email [drop the address] to receive instructions on resetting your password"..
- The button should read sth along the lines of "Send password reset instruction" or "Send instructions"
- The "Back to Akvo" link at the bottom of the screen is a tiny bit misleading, the user hasn't left Akvo. I think that should read "Login" since it actually takes the user to the Login screen

**Comment by Kiarii** [comment]
#### Password reset instructions sent 
- The title to that screen should read "Password reset instructions sent"
- Not sure if the desc text is necessary in this case..
- We do not check email validity on this action :( notice my email ends with a **.or**. This we must of course do, and tell the user if the email they provide is not registered to an Akvo account. This should happen in the previous screen, and the error msg should also be shown there..
<img width="406" alt="Screenshot 2019-10-03 at 12 26 39" src="https://user-images.githubusercontent.com/12692444/66115278-2c92e880-e5d9-11e9-8591-3c00edfe567b.png">

**Comment by janagombitova** [comment]
@Kiarii I am going to close this issue as for a 1st release all works well and comment on changes we will **not** make and why. For the other changes and suggestions, I am creating a new issue I will link here. I also realise I should have briefed you a bit better on what we can and what we cannot change. Better next time. 

Everything in this comment (https://github.com/akvo/akvo-platform/issues/154#issuecomment-537841147) about the initial log in change is something we will not change. We have decided to move to Auth0 and use what they provide to not have to handle login on our own and make use of their service. This page is what they provide and we can only change the colour and the logo. 

When it comes to the overall terminology (https://github.com/akvo/akvo-platform/issues/154#issuecomment-537853055) we will do a round with MarComms later in the future to update the copy of the email and related pages to make it sound more Akvo. I will create an issue for later. When it comes to giving the user a url to bring them to Flow/Lumen after they verified/activated their account this is currently not possible due to our set up. We do not want to be too dependent on A0uth0 so it has no way to know which Flow/Lumen instance to send the user to. 

Password reset (https://github.com/akvo/akvo-platform/issues/154#issuecomment-537861006 and https://github.com/akvo/akvo-platform/issues/154#issuecomment-537866544) are more pages provided by Auth0 we do not have control over as we do not want to have to create all these workflows on our own. We use what we get. 

**Small possible improvements (in separate issues):** 
* improve the behaviour of user profile pop up 
* change 'Yikes' in the email verification page 
* terminology improvements in collaboration with MarComms

**Comment by Kiarii** [comment]
#### Pwd reset email
Here as with the account verification we create terminology discrepancies, we go from "password reset" to "password change". Note that "password change" or more appropriately "Change password" is an action the user is likely to do while they are logged in into their account, not something that they order via an email. That said 
- the email should be titled "Reset password" 
- the first part of the desc shd read "Click on the button below to reset your password"
- the link/button: "Reset password"
- the latter desc: "If you did not request ..."
<img width="600" alt="Screenshot 2019-10-03 at 12 54 02" src="https://user-images.githubusercontent.com/12692444/66117787-d1afc000-e5dd-11e9-923c-8b19413739b3.png">

**Comment by Kiarii** [comment]
ok, no more comments then :)
